### ISSUE TITLE
Plugin selection based on node's capability

### ISSUE THREAD

**Issue by madejackson** [HaveAGitGat/Tdarr#504]
If I have multiple nodes working on 1 library, I'd like to select the Plugin based on the node's HW-Capability. This would allow me to use NVENC and QVS nodes simultaniously.

The way plugins are processed right now, I see no possible way to do this other than using multiple libraries.

**Comment by HaveAGitGat** [comment]
You can set the node hardware type in the node options panel on the Tdarr tab.

So keep both qsv and nvenc plugins checked in the library settings, and then create an nvenc and qsv node. The nodes should then auto detect which plugins they can do.

**Comment by madejackson** [comment]
While this is true, it does not work for my use case.

Use case: Plugin order NVENC->QVS
Every file is processed by every node once. QVS-capable node is waiting for NVENC-node to process the file and then tries to process it, find it's already processed to HVEC and hence does nothing. So my QVS-node is idling all the time doing none of the work.

I'd like to process every file once from one node only and the other Plugin should be skipped/omitted.

It's possible that I am missing something but that's what I experienced and what I understand of this:
![example_plugin_stack_flow](https://user-images.githubusercontent.com/24832654/141340206-2cd75560-e901-44e9-b616-b0a8c12878a1.png)

**Comment by HaveAGitGat** [comment]
Btw it’s ‘QSV’ and sorry I’m not too sure what you mean.

Like I said you can achieve what you’re after by having an nvenc and qsv node (you can have multiple nodes on the same machine in case your machine is both qsv and nvenc capable). This allows 2 files to be processed in parallel (start a worker on each node), one with nvenc and one with qsv. It’s not possible to process a single file with nvenc and qsv at the same time. 

So my suggestion is:
- Add a qsv and and nvenc plugin to your plug-in stack.
- Create a node and then on the node options panel on the ‘Tdarr’ tab, select the hardware type as qsv.
-

**Comment by HaveAGitGat** [comment]
Woops didn’t mean to close, fat fingers hit the wrong button.

**Comment by brayrobert201** [comment]
Expanding this out a bit:

I have a node with a GPU capable of transcoding. I've got a computer that would be a potential node that's CPU only. I'd love to have whatever device act based on what's actually available.

The same thing could also be done for unused CPUs.

IE: If GPU available, then use Tdarr_Plugin_d5d3_iiDrakeii_FFMPEG_NVENC_Tiered_MKV
Else use Tdarr_Plugin_MC93_Migz1FFMPEG_CPU

The CPUs aren't as fast as the GPUs, obviously, but it's still unused resources that could be used for other purposes.

**Comment by CyberSardinha** [comment]
> Expanding this out a bit:
> 
> I have a node with a GPU capable of transcoding. I've got a computer that would be a potential node that's CPU only. I'd love to have whatever device act based on what's actually available.
> 
> The same thing could also be done for unused CPUs.
> 
> IE: If GPU available, then use Tdarr_Plugin_d5d3_iiDrakeii_FFMPEG_NVENC_Tiered_MKV Else use Tdarr_Plugin_MC93_Migz1FFMPEG_CPU
> 
> The CPUs aren't as fast as the GPUs, obviously, but it's still unused resources that could be used for other purposes.

This would be amazing! we could use the full power of the nodes. as it stands at the moment my cpu is most of the time doing noting because all the tasks are waiting for a gpu. i would like to encode using software if all the gpu slots are in use as you said.

**Comment by HaveAGitGat** [comment]
>I have a node with a GPU capable of transcoding. I've got a computer that would be a potential node that's CPU only. I'd love to have whatever device act based on what's actually available.

If this is the case then only start GPU workers on the GPU node and only start CPU workers on the CPU only node.

For example, in my plugin stack I have a GPU and CPU plugin:

![image](https://user-images.githubusercontent.com/43864057/155079822-d2a8c78d-a53e-4d0b-bf31-ba4bb9d346df.png)

I start a CPU and a GPU worker:

![image](https://user-images.githubusercontent.com/43864057/155079854-daf2603a-0839-4b61-bcd3-0e39540329dd.png)

The CPU worker uses the CPU plugin and the GPU worker uses the GPU plugin, so both CPU and GPU are highly utilized:
![image](https://user-images.githubusercontent.com/43864057/155079943-25a87552-14cf-4474-beb0-827c7da2fd4c.png)
![image](https://user-images.githubusercontent.com/43864057/155079925-1ce99f9d-df51-421c-9de1-d8394131bad0.png)

Make sure to turn off `Allow GPU workers to do CPU tasks` in the Node options so the GPU worker doesn't do the CPU task.

**Comment by CyberSardinha** [comment]
If I do as you explained and organize the stack like that only the cpu gets used.

**Comment by Starker3** [comment]
Is it possible to do this same sort of setup for two nodes that both have GPU's?
I have one QSV node and one NVENC node. When I set the QSV node to 1 GPU worker with an NVENC plugin in the stack it doesn't process the file and just adds the following to the log:

`Node[NodeName]:Worker[dRBFzM-fi]:[Step W03.1] [C2] Won't process: requireGPU:nvenc`

Basically I want the QSV node to use the QSV specific plugins when processing and skip using the NVENC plugin and I want the NVENC node to skip using the QSV plugin and only use the NVENC plugin.

**Comment by supersnellehenk** [comment]
> Basically I want the QSV node to use the QSV specific plugins when processing and skip using the NVENC plugin and I want the NVENC node to skip using the QSV plugin and only use the NVENC plugin.

You can set the capabilities of the node in the node options. Then it'll do as you've described.

**Comment by Starker3** [comment]
> > Basically I want the QSV node to use the QSV specific plugins when processing and skip using the NVENC plugin and I want the NVENC node to skip using the QSV plugin and only use the NVENC plugin.
> 
> You can set the capabilities of the node in the node options. Then it'll do as you've described.

This is what I tried. But setting a specific node to use QSV only when there is an NVENC plugin will not let that node process the file. That's the log output I pasted when I try it:
`Node[NodeName]:Worker[dRBFzM-fi]:[Step W03.1] [C2] Won't process: requireGPU:nvenc`

**Comment by supersnellehenk** [comment]
> This is what I tried. But setting a specific node to use QSV only when there is an NVENC plugin will not let that node process the file. That's the log output I pasted when I try it:
> 
> `Node[NodeName]:Worker[dRBFzM-fi]:[Step W03.1] [C2] Won't process: requireGPU:nvenc`

In that case I'd try opening a new issue instead of commenting on this one, it'll get seen earlier than a closed issue. Should work how you've set it up as far as I know.

**Comment by Starker3** [comment]
I'll check again to see if there is anything I've setup incorrectly and try again. If I still don't get it right I'll create a new issue. I commented here since it was similar OPs situation.

**Comment by Blackpaw** [comment]
Jumping in (necrothreading?) to say I'm seeing this problem too - have two GPU Nodes
* VAAPI
*  NVENC
Both marked as vaapi & nvenc only in the respective node options. I have vaapi and nvenc plugins set in the processing stack.

But when ever the vaapi node gets sent a file it just gets marked as "requires NVENC worker"

  `Node[InternalVAAPINode]:Worker[H5VrumBSx]:[Step W03.1] [C1] Won't process: requireGPU:nvenc`

Nodes and server are latest versions.

**Comment by Jinra** [comment]
@Blackpaw did you ever find a solution for this? I'm running into the exact same issue where only nodes with their respective encoder plugin are higher in priority for the stack are running and others will just say it requires a different GPU.

If I switch the priority around the exactly opposite will happen to the other workers.

**Comment by HaveAGitGat** [comment]
@Jinra make sure to disable this option `Process Plugins Sequentially`:
![image](https://github.com/HaveAGitGat/Tdarr/assets/43864057/312ef4ee-c3a0-4a77-a23c-bd7ff68e7abe)

Will allow the worker to jump ahead to the plugin that it can do.

**Comment by Du-Kee** [comment]
I am also having this issue. I have unchecked "Process Plugins Sequentially", and I have the QSV and NVENC nodes appropriately selected on the tdarr tab. If I rearrange the plugins then it just flips it the other way and the second node in the plugin sequence fails the job with either the error `Won't process: requireGPU:qsv` or `Won't process: requireGPU:nvenc`

I've attempted it with several different plugins, and either will work if the other is unchecked. It seems determined to process them in order.

**Comment by cameronkelley** [comment]
@Du-Kee try checking and then unchecking it. Someone asked a while ago on reddit and I thinks it's a UI issue.

https://www.reddit.com/r/Tdarr/comments/180j9xh/gpu_worker_required/

@HaveAGitGat forgot to mention this issue when I found it, but might still be a problem.

**Comment by Du-Kee** [comment]
@cameronkelley Thanks for the response! 

I tried toggling it, and also restarted both the server and the nodes. The server is UnRAID (via Docker) and the nodes are on a Windows laptop running Windows 11, with 11800H CPU and 3050ti dGPU. 

Different but adjacent issue: It actually did work for a short time last night, until I had to reboot the PC containing the nodes for something unrelated. After that, I couldn't get the iGPU to work at all using QSV ("GPU worker needed" was the message in the staging area) until I stuck a newer version of ffmpeg in the ffmpeg-static folder. The same error was still occurring prior to the ffmpeg upgrade, it just wasn't working on the QSV side of things. The logs reported that no GPU was present in those cases. 

After the ffmpeg update, I confirmed that QSV was working with the iGPU, and separately confirmed that nvenc was working with the dGPU, isolating both by unchecking their associated plugins in the stack. 

I'm more than happy to do some testing to help isolate the issue. It seems a lot of people have run into this over a long period of time. I think tdarr is amazing and I'd love to contribute!

**Comment by dennix85** [comment]
Well if you are willing to swap the stack for flows you can do this by the check node hardware encoder block, set 2 of the right after each other, one to check nvenc, ont to check qsv, and let them go the the plugin you want that hardware to process..
